File: split-repos/kodama-r/inst/benchmarks/run_br8100_merfish.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages(library(kodamaR))

get_env <- function(name, default) {
  value <- Sys.getenv(name, "")
  if (nzchar(value)) value else default
}

first_existing <- function(paths) {
  hit <- paths[file.exists(paths)][1]
  if (length(hit) == 0L || is.na(hit)) paths[1] else hit
}

load_spatial_dataset <- function(path) {
  env <- new.env(parent = emptyenv())
  load(path, envir = env)
  if (exists("dataset", envir = env, inherits = FALSE)) {
    ds <- get("dataset", envir = env)
    return(list(data = as.matrix(ds$data), spatial = as.matrix(ds$spatial), labels = as.factor(ds$labels)))
  }
  names <- ls(env)
  data_name <- intersect(c("pca.PM", "pca_Br8100", "data", "x"), names)[1]
  spatial_name <- intersect(c("xyz", "xy_Br8100", "spatial"), names)[1]
  labels_name <- intersect(c("tissue_segments", "labels_Br8100", "labels"), names)[1]
  if (any(is.na(c(data_name, spatial_name, labels_name)))) {
    stop("Cannot identify data/spatial/labels objects in ", path)
  }
  list(
    data = as.matrix(get(data_name, envir = env)),
    spatial = as.matrix(get(spatial_name, envir = env)),
    labels = as.factor(get(labels_name, envir = env))
  )
}

run_one <- function(name, ds, classifier, backend, M, Tcycle, landmarks, knn.k, ncomp) {
  message(sprintf("[%s] %s backend=%s M=%d Tcycle=%d", name, classifier, backend, M, Tcycle))
  t0 <- proc.time()[["elapsed"]]
  result <- kodama_matrix(
    ds$data,
    spatial = ds$spatial,
    M = M,
    Tcycle = Tcycle,
    ncomp = ncomp,
    landmarks = landmarks,
    knn.k = knn.k,
    classifier = classifier,
    backend = backend,
    progress = TRUE
  )
  elapsed <- proc.time()[["elapsed"]] - t0
  classes <- apply(result$res, 1L, function(z) length(unique(z)))
  data.frame(
    dataset = name,
    classifier = classifier,
    backend = backend,
    samples = nrow(ds$data),
    variables = ncol(ds$data),
    M = M,
    Tcycle = Tcycle,
    landmarks = landmarks,
    knn_k = knn.k,
    ncomp = ncomp,
    runtime_seconds = elapsed,
    core_runtime_seconds = result$runtime_seconds,
    median_acc = median(result$acc),
    max_acc = max(result$acc),
    median_classes = median(classes),
    min_classes = min(classes),
    max_classes = max(classes)
  )
}

out_dir <- get_env("KODAMA_R_BENCH_OUT", file.path(getwd(), "kodama-r-benchmarks"))
dir.create(out_dir, recursive = TRUE, showWarnings = FALSE)

datasets <- list(
  MERFISH = get_env("MERFISH_RDATA", first_existing(c(
    file.path("data", "MERFISH.RData"),
    file.path("data", "MERFISH-input.RData")
  ))),
  Br8100 = get_env("BR8100_RDATA", first_existing(c(
    file.path("data", "DLFPC-Br8100.RData"),
    file.path("data", "DLFPC-Br8100-input.RData"),
    file.path("data", "Br8100.RData")
  )))
)

M <- as.integer(get_env("KODAMA_BENCH_M", "20"))
Tcycle <- as.integer(get_env("KODAMA_BENCH_TCYCLE", "20"))
landmarks <- as.integer(get_env("KODAMA_BENCH_LANDMARKS", "100000"))
knn.k <- as.integer(get_env("KODAMA_BENCH_KNN_K", "30"))
ncomp <- as.integer(get_env("KODAMA_BENCH_NCOMP", "50"))
backends <- strsplit(get_env("KODAMA_BENCH_BACKENDS", "cpu,cuda"), ",", fixed = TRUE)[[1]]

rows <- list()
for (dataset_name in names(datasets)) {
  path <- datasets[[dataset_name]]
  if (!file.exists(path)) {
    warning("Skipping missing dataset: ", path)
    next
  }
  ds <- load_spatial_dataset(path)
  for (backend in backends) {
    for (classifier in c("knn", "pls_lda")) {
      rows[[length(rows) + 1L]] <- run_one(dataset_name, ds, classifier, backend, M, Tcycle, landmarks, knn.k, ncomp)
      write.csv(do.call(rbind, rows), file.path(out_dir, "br8100_merfish_summary.csv"), row.names = FALSE)
    }
  }
}

print(do.call(rbind, rows))

File: split-repos/kodama-python/benchmarks/run_br8100_merfish.py
#!/usr/bin/env python3
import argparse
import csv
import time
from pathlib import Path

import numpy as np

import kodama


def load_csv_matrix(path: Path) -> np.ndarray:
    return np.loadtxt(path, delimiter=",", skiprows=1, dtype=np.float32)


def load_dataset(input_dir: Path, name: str):
    data = load_csv_matrix(input_dir / f"{name}_data.csv.gz")
    spatial = load_csv_matrix(input_dir / f"{name}_spatial.csv.gz")
    return data, spatial


def run_one(name, data, spatial, classifier, backend, args):
    print(f"[{name}] {classifier} backend={backend} M={args.M} Tcycle={args.Tcycle}", flush=True)
    start = time.perf_counter()
    result = kodama.matrix(
        data,
        spatial=spatial,
        M=args.M,
        Tcycle=args.Tcycle,
        ncomp=args.ncomp,
        landmarks=args.landmarks,
        knn_k=args.knn_k,
        classifier=classifier,
        backend=backend,
        progress=True,
    )
    elapsed = time.perf_counter() - start
    classes = np.array([len(np.unique(row)) for row in result["res"]])
    return {
        "dataset": name,
        "classifier": classifier,
        "backend": backend,
        "samples": data.shape[0],
        "variables": data.shape[1],
        "M": args.M,
        "Tcycle": args.Tcycle,
        "landmarks": args.landmarks,
        "knn_k": args.knn_k,
        "ncomp": args.ncomp,
        "runtime_seconds": elapsed,
        "core_runtime_seconds": float(result["runtime_seconds"]),
        "median_acc": float(np.median(result["acc"])),
        "max_acc": float(np.max(result["acc"])),
        "median_classes": float(np.median(classes)),
        "min_classes": int(np.min(classes)),
        "max_classes": int(np.max(classes)),
    }


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--input-dir", default="exported-spatial")
    parser.add_argument("--out", default="kodama-python-br8100-merfish.csv")
    parser.add_argument("--M", type=int, default=20)
    parser.add_argument("--Tcycle", type=int, default=20)
    parser.add_argument("--landmarks", type=int, default=100000)
    parser.add_argument("--knn-k", type=int, default=30)
    parser.add_argument("--ncomp", type=int, default=50)
    parser.add_argument("--backends", default="cpu,cuda")
    args = parser.parse_args()

    input_dir = Path(args.input_dir)
    rows = []
    for name in ("MERFISH", "Br8100"):
      data, spatial = load_dataset(input_dir, name)
      for backend in args.backends.split(","):
        for classifier in ("knn", "pls_lda"):
          row = run_one(name, data, spatial, classifier, backend, args)
          rows.append(row)
          with open(args.out, "w", newline="") as handle:
            writer = csv.DictWriter(handle, fieldnames=list(rows[0].keys()))
            writer.writeheader()
            writer.writerows(rows)
    for row in rows:
        print(row)


if __name__ == "__main__":
    main()
